### PR TITLE
feat(sync-service): Support missing comparison predicates

### PR DESCRIPTION
**Pull request by robacourt**
Support for the missing comparison predicates in WHERE clauses, as mentioned in #1441:
-  x BETWEEN y AND z
-  x NOT BETWEEN y AND z
-  z BETWEEN SYMMETRIC y AND z
-  z NOT BETWEEN SYMMETRIC y AND z
-  x IS UNKNOWN
-  x IS NOT UNKNOWN

### CODE REVIEW

File: packages/sync-service/lib/electric/replication/eval/env/explicit_casts.ex [hunk 1]
```diff
@@ -6,49 +6,51 @@ defmodule Electric.Replication.Eval.Env.ExplicitCasts do
...
+  | jsonb     | float8         |                 |
+  | jsonb     | numeric        |                 |
+  | int4range | int4multirange |                 |
+  | numrange  | nummultirange  |                 |
+  | tsrange   | tsmultirange   |                 |
+  | tstzrange | tstzmultirange |                 |
+  | daterange | datemultirange |                 |
+  | int8range | int8multirange |                 |
+  | xid8      | xid            |                 |
+  | unknown   | bool           | unknown_to_bool |
```
Comment (icehaunter) [inline packages/sync-service/lib/electric/replication/eval/env/explicit_casts.ex:46]: Why do we need that explicit cast? I believe that an explicit `::bool` cast shouldn't be special-cased here in this list, but instead should always work for a `NULL` value, as it always tries to work for text values

File: packages/sync-service/lib/electric/replication/eval/parser.ex [hunk 2]
```diff
@@ -365,26 +372,68 @@ defmodule Electric.Replication.Eval.Parser do
...
+         comparisons = [left_comparison, right_comparison],
+         {:ok, comparisons} <- Utils.map_while_ok(comparisons, &maybe_reduce/1),
+         {:ok, reduced} <-
+           build_bool_chain(%{name: "and", impl: &Kernel.and/2}, comparisons, expr.location) do
+      {:ok, reduced}
     end
   end
 
-  defp build_or_chain([head | tail], location) do
+  defp between_sym(expr, left_bound, right_bound, refs, env) do
```
Comment (icehaunter) [inline packages/sync-service/lib/electric/replication/eval/parser.ex:421]: Can you add a comment here about suboptimality of recalculating a subtree for both branches? We don't have to address it today, just for the futute

**Review by icehaunter — CHANGES_REQUESTED** [review]
LGTM, with one request about explicit casts

File: packages/sync-service/lib/electric/replication/eval/parser.ex (hunk 2 above)
Comment (robacourt) [reply packages/sync-service/lib/electric/replication/eval/parser.ex:421]: Out of curiosity, how would it be optimised? I thought of doing a seeing whether the left or right bounds were greater before doing the between but I'm didn't think that would be any quicker apart from in specific cases such as left bound and right bound being constants.

File: packages/sync-service/lib/electric/replication/eval/env/explicit_casts.ex (hunk 1 above)
Comment (robacourt) [reply packages/sync-service/lib/electric/replication/eval/env/explicit_casts.ex]: Something like this? https://github.com/electric-sql/electric/pull/1569/commits/d18be4af952bd884ee8c20284e3e424532b43e0e

File: packages/sync-service/lib/electric/replication/eval/parser.ex (hunk 2 above)
Comment (robacourt) [reply packages/sync-service/lib/electric/replication/eval/parser.ex:422]: Done.

**Pull request merged**